devicetypes/plaidsystems/spruce-controller.src/spruce-controller.groovy

```diff
@@ -14,736 +10,525 @@
+
+//get times from settings and send to controller, then start schedule
+def startSchedule() {
+	def startRun = false
+	def runTime, totalTime=0
+	def scheduleTimes = []
+
+	for (i in 1..16) {
+		def endpoint = i + 1
+		//if (settings."z${i}" && settings."z${i}Duration" != null) {
+		if (settings."z${i}Duration" != null) {
+			runTime = Integer.parseInt(settings."z${i}Duration")
+			totalTime += runTime
+			startRun = true
+
+			scheduleTimes.push(zigbee.writeAttribute(zigbee.ONOFF_CLUSTER, OFF_WAIT_TIME_ATTRIBUTE, DataType.UINT16, runTime, [destEndpoint: endpoint]))
+		}
+		else {
+			scheduleTimes.push(zigbee.writeAttribute(zigbee.ONOFF_CLUSTER, OFF_WAIT_TIME_ATTRIBUTE, DataType.UINT16, 0, [destEndpoint: endpoint]))
+		}
+	}
+	if (!startRun || totalTime == 0) return noSchedule()
+
+	//start after scheduleTimes are sent
+	scheduleTimes.push(zigbee.command(zigbee.ONOFF_CLUSTER, 1, "", [destEndpoint: 1]))
+	sendEvent(name: "status", value: "Scheduled for ${totalTime}min(s)", descriptionText: "Start schedule ending in ${totalTime} mins")
+	return scheduleTimes
 }
-def z14off() {
-    "st cmd 0x${device.deviceNetworkId} 15 6 0 {}"
+
+//write switch time settings map
+def settingsMap(WriteTimes, attrType) {
+
+	def runTime
+	def sendCmds = []
+	for (endpoint in 1..17) {
+		if (WriteTimes."${endpoint}") {
+			runTime = Integer.parseInt(WriteTimes."${endpoint}")
+
+			if (attrType == ON_TIME_ATTRIBUTE) sendCmds.push(zigbee.writeAttribute(zigbee.ONOFF_CLUSTER, ON_TIME_ATTRIBUTE, DataType.UINT16, runTime, [destEndpoint: endpoint]))
+			else sendCmds.push(zigbee.writeAttribute(zigbee.ONOFF_CLUSTER, OFF_WAIT_TIME_ATTRIBUTE, DataType.UINT16, runTime, [destEndpoint: endpoint]))
+		}
+	}
+	return sendCmds
 }
-def z15on() {
-	return manual() + "st cmd 0x${device.deviceNetworkId} 16 6 1 {}"  
+
+//send switch time
+def writeType(endpoint, cycle) {
+	zigbee.writeAttribute(zigbee.ONOFF_CLUSTER, ON_TIME_ATTRIBUTE, DataType.UINT16, cycle, [destEndpoint: endpoint])
 }
-def z15off() {
-	"st cmd 0x${device.deviceNetworkId} 16 6 0 {}"    
+
+//send switch off time
+def writeTime(endpoint, runTime) {
+	zigbee.writeAttribute(zigbee.ONOFF_CLUSTER, OFF_WAIT_TIME_ATTRIBUTE, DataType.UINT16, runTime, [destEndpoint: endpoint])
 }
-def z16on() {
-	return manual() + "st cmd 0x${device.deviceNetworkId} 17 6 1 {}"    
+
+//set reporting and binding
+def configure() {
+	// Device-Watch checks every 1 hour
+	sendEvent(name: "checkInterval", value: HC_INTERVAL_MINS * 60, displayed: false, data: [protocol: "zigbee", hubHardwareId: device.hub.hardwareID])
+
+	if (DEBUG) log.debug "Configuring Reporting and Bindings ${device.name} ${device.deviceNetworkId} ${device.hub.zigbeeId}"
+
+	//setup binding for 18 endpoints
+	def bindCmds = []
+	bindCmds += zigbee.addBinding(zigbee.ONOFF_CLUSTER, [destEndpoint: 1])
+	bindCmds += zigbee.addBinding(ALARMS_CLUSTER, [destEndpoint: 1])
+
+	for (endpoint in 1..18) {
+		bindCmds += zigbee.addBinding(BINARY_INPUT_CLUSTER, [destEndpoint: endpoint])
```

Comment:
Just a note re the comment below: https://github.com/SmartThingsCommunity/SmartThingsPublic/pull/47339/files?diff=unified&w=1#r518333854

The `zigbee.configureReporting` internally calls `addBinding(cluster, additionalParams)`, so these packets will be sent twice. `configure()` isn't something that happens often, but there is always the chance that the extra packets could tie things up or get dropped (especially if the link to the device is weak), or cause issues for the Zigbee network if there are many devices on it.

Reply:
Ok, removed bindCmds and associated code, since reporting handles this